### ISSUE TITLE
Don't follow source maps in files when they contain the sources itself / doesn't work with embedded sources in source map

### ISSUE THREAD

**Issue by phip1611** [systemjs/builder#787]
Hi,
I compiled typescript to javascript with the compiler options `inlineSourceMap` and `inlineSources`. The compiled files look like this:

```
...
], AppModule);
exports.AppModule = AppModule;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjo... long base 64 string
```

decoded base64 decoded looks like this:
```
{
    "version":3,
    "file":"app.module.js",
    "sourceRoot":"",
    "sources":[
        "app.module.ts"
    ],
    "names":[

    ],
    "mappings":";;;;;;;AAAA,+DAA4D;...",
    "sourcesContent":[
        "import { ErrorPageComponent } from './error-page.component';\ni..."
    ]
}
```

**The Bug**
SystemJS-Builder seems to ignore that there is a `sourcesContent`-field. Instead it sees the `sources` field and replaces the source-url **in the bundle source map** with this typescript file **instead** the original javascript file. The typescript files aren't deployed on my server. 

**Expected behaviour**
SystemJS-Bundle-Map should point to the javascript files and **don't** follow their source maps. Or, maybe even better, it should take these sourcesContent-value and put it into the bundle-source-map.

PS: I have to admit that for now it seems that anyway no browser can follow source maps for multiple steps. But maybe in the future.

**Comment by guybedford** [comment]
Agreed this should be the behaviour for sourcesContents when processing source maps here.

**Comment by phip1611** [comment]
I think the best solution would be to use the sourcesContents from the javascript-files and to embed it into the bundles source map, yes.

PS: The other option to point to the javascript files which themselves point to their own typescript source wouldn't work because for now no browser can follow source maps for more than one step, as I mentioned above. **But** I contacted the chromium developers to implement following source maps for more than one step. They assigned the bug/feature request. Perhaps both solutions might work in the future. https://bugs.chromium.org/p/chromium/issues/detail?id=704120&q=&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified&start=200

**Comment by thomas-darling** [comment]
I'm somewhat confused about the behavior you seem to suggest here...

We **absolutely** need to be able to map a bundle back to it's original TypeScript code, and we **absolutely** need the option to do that by including the original TypeScript source as `sourceContent` in the source map for the bundle.

Anything else would make debugging production code a complete nightmare, and it would significantly limit the usefulness of e.g. the various cloud-based error reporting services, which use source maps to map stack traces back to the original source for easier debugging.

Now, of course you don't want to serve your source maps to the public, as that would reveal your original code - but that's why such error reporting services allow you to explicitly upload your source maps when you make a release, or protect them by using authentication when they are fetched from your server.

I actually just made a related pull request, which fixes a path-normalization bug that prevented the original source from being included in some scenarios: https://github.com/systemjs/builder/pull/823
Once this is merged, the source map for the bundle will contain mappings and source for _both_ the original TypeScript files, _and_ the transpiled JavaScript files being bundled. This is how it should be, as it allows you to debug either the original source, or the transpiled source, depending on your needs.

An option to choose whether to include the original or transpiled source in the bundle source map could also be acceptable - but I must stress, that the option to include the TypeScript source as `sourceContent` in the source map for the bundle is definitely a requirement.
